service/src/main/java/com/codeforcommunity/processor/ProtectedSiteProcessorImpl.java

```diff
@@ -0,0 +1,247 @@
+package com.codeforcommunity.processor;
+
+import static org.jooq.generated.Tables.ADOPTED_SITES;
+import static org.jooq.generated.Tables.SITES;
+import static org.jooq.generated.Tables.SITE_ENTRIES;
+import static org.jooq.generated.Tables.STEWARDSHIP;
+
+import com.codeforcommunity.api.IProtectedSiteProcessor;
+import com.codeforcommunity.auth.JWTData;
+import com.codeforcommunity.dto.site.AddSiteRequest;
+import com.codeforcommunity.dto.site.AdoptedSitesResponse;
+import com.codeforcommunity.dto.site.RecordStewardshipRequest;
+import com.codeforcommunity.dto.site.UpdateSiteRequest;
+import com.codeforcommunity.enums.PrivilegeLevel;
+import com.codeforcommunity.exceptions.AuthException;
+import com.codeforcommunity.exceptions.ResourceDoesNotExistException;
+import com.codeforcommunity.exceptions.WrongAdoptionStatusException;
+import java.sql.Timestamp;
+import java.util.List;
+import org.jooq.DSLContext;
+import org.jooq.generated.tables.records.AdoptedSitesRecord;
+import org.jooq.generated.tables.records.SiteEntriesRecord;
+import org.jooq.generated.tables.records.SitesRecord;
+import org.jooq.generated.tables.records.StewardshipRecord;
+
+public class ProtectedSiteProcessorImpl implements IProtectedSiteProcessor {
+
+  private final DSLContext db;
+
+  public ProtectedSiteProcessorImpl(DSLContext db) {
+    this.db = db;
+  }
+
+  private void checkSiteExists(int siteId) {
+    if (!db.fetchExists(db.selectFrom(SITES).where(SITES.ID.eq(siteId)))) {
+      throw new ResourceDoesNotExistException(siteId, "Site");
+    }
+  }
+
+  private Boolean isAlreadyAdopted(int userId, int siteId) {
+    return db.fetchExists(
+        db.selectFrom(ADOPTED_SITES)
+            .where(ADOPTED_SITES.USER_ID.eq(userId))
+            .and(ADOPTED_SITES.SITE_ID.eq(siteId)));
+  }
+
+  /**
+   * Throws an exception if the user is not an admin or super admin.
+   *
+   * @param level the privilege level of the user calling the route
+   */
+  void isAdminCheck(PrivilegeLevel level) {
+    if (!(level.equals(PrivilegeLevel.ADMIN) || level.equals(PrivilegeLevel.SUPER_ADMIN))) {
+      throw new AuthException("User does not have the required privilege level.");
+    }
+  }
+
+  @Override
+  public void adoptSite(JWTData userData, int siteId) {
+    checkSiteExists(siteId);
+    if (isAlreadyAdopted(userData.getUserId(), siteId)) {
+      throw new WrongAdoptionStatusException(true);
+    }
+
+    AdoptedSitesRecord record = db.newRecord(ADOPTED_SITES);
+    record.setUserId(userData.getUserId());
+    record.setSiteId(siteId);
+    record.store();
+  }
+
+  @Override
+  public void unadoptSite(JWTData userData, int siteId) {
+    checkSiteExists(siteId);
+    if (!isAlreadyAdopted(userData.getUserId(), siteId)) {
+      throw new WrongAdoptionStatusException(false);
+    }
+
+    db.deleteFrom(ADOPTED_SITES)
```

Comment:
I haven't seen the schema for this db in a while, but just to double check, are we not doing soft deletes on this table?

Reply:
I'm not sure, @florisdobber @oloveluck do either of you have thoughts on this?